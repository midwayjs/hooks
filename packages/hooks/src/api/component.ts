import { IMidwayContainer } from '@midwayjs/core'
import { ApplicationContext, Configuration, Init } from '@midwayjs/decorator'
import { __decorate } from 'tslib'
import {
  AbstractRouter,
  ApiRouter,
  createDebug,
  useFrameworkAdapter,
  validateArray,
} from '@midwayjs/hooks-core'
import {
  getHydrateOptions,
  getRouter,
  getSource,
  HOOKS_DEV_MODULE_PATH,
  isDev,
  isHydrate,
  loadApiFiles,
  parseApiModule,
  RuntimeConfig,
} from '@midwayjs/hooks-internal'
import flattenDeep from 'lodash/flattenDeep'
import { MidwayFrameworkAdapter } from './component/adapter'

const debug = createDebug('hooks: component')

export function HooksComponent(runtimeConfig: RuntimeConfig = {}) {
  if (runtimeConfig.middleware !== undefined) {
    validateArray(runtimeConfig.middleware, 'runtimeConfig.middleware')
  }

  const useSourceFile = isDev()
  debug('useSourceFile: %s', useSourceFile)

  const source = getSource({ useSourceFile })
  const router = getRouter(source)
  const midway = new MidwayFrameworkAdapter(source, router, null)

  useFrameworkAdapter({
    name: 'Midway',
    handleResponseMetaData: midway.handleResponseMetaData,
  })

  const apis = isHydrate()
    ? loadHydrateModules(router)
    : loadApiModules(source, router)

  if (apis.length === 0) {
    console.warn('No api routes found, source is:', source)
  }

  debug('HOOKS_DEV_MODULE_PATH', process.env[HOOKS_DEV_MODULE_PATH])
  if (process.env[HOOKS_DEV_MODULE_PATH]) {
    const devApis = loadApiModules(
      process.env[HOOKS_DEV_MODULE_PATH],
      router instanceof ApiRouter ? router : new ApiRouter()
    )
    debug('load dev apis %O', devApis)
    apis.push(...devApis)
  }

  const Configuration = createConfiguration(async (container) => {
    await midway.initialize(container, apis, runtimeConfig.middleware)
  })

  return { Configuration }
}

// source: https://www.typescriptlang.org/play?target=99&moduleResolution=99&importHelpers=true&emitDecoratorMetadata=false&pretty=true#code/JYWwDg9gTgLgBAbzgSQHbBgGjgQTGAG2AGMBDGYCVAYSpgFMAPLOW1AM2AHMBXKcyqjgBfOOygQQcAEQABEMAAmAd1IBPAFYBnAPSL6xaOWjSA3AChL52W069+FKgAoE5uHFSkQ9LWFLF6AC4ZeSVVTV0ACwgIAGstaUxzYQBKc2ICUi0tOAAJGPjbbgdBRDc4WTxCEgEqNgZmJzT3Q1QYUmBUeihg0lQ1C3LZNAwm8qy1VGI4TtGUsvdFuFJVDDgqEZgnGEjgLQA6VvbO7uaRZPMgA
function createConfiguration(
  onInit: (container: IMidwayContainer) => Promise<void>
) {
  let HooksConfiguration = class HooksConfiguration {
    container: any
    async init() {
      await onInit(this.container)
    }
  }
  __decorate(
    [ApplicationContext()],
    HooksConfiguration.prototype,
    'container',
    void 0
  )
  __decorate([Init()], HooksConfiguration.prototype, 'init', null)
  HooksConfiguration = __decorate(
    [
      Configuration({
        namespace: '@midwayjs/hooks',
        importConfigs: [
          {
            default: {
              asyncContextManager: {
                enable: true,
              },
            },
          },
        ],
      }),
    ],
    HooksConfiguration
  )

  return HooksConfiguration
}

function loadHydrateModules(router: AbstractRouter) {
  const { modules } = getHydrateOptions()
  debug('load hydrate modules %O', modules)
  const routes = modules.map(({ mod, file }) =>
    parseApiModule(mod, file, router)
  )
  return flattenDeep(routes)
}

function loadApiModules(source: string, router: AbstractRouter) {
  const { apis } = loadApiFiles({
    source,
    router,
  })
  const routes = apis.map((file) => parseApiModule(require(file), file, router))
  return flattenDeep(routes)
}
