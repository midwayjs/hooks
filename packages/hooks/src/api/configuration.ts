import isFunction from 'lodash/isFunction'
import { createConfiguration as createConfigurationFromCore } from '@midwayjs/core'
import { InjectionConfigurationOptions } from '@midwayjs/decorator'

type FunctionalConfiguration = ReturnType<typeof createConfigurationFromCore>
type FirstArgument<T> = T extends (arg: infer A, ...args: any[]) => any
  ? A
  : never
type ConfigurationLifeCycle = FirstArgument<FunctionalConfiguration['onReady']>

interface ConfigurationOptions extends InjectionConfigurationOptions {
  onReady?: ConfigurationLifeCycle
  onStop?: ConfigurationLifeCycle
  onConfigLoad?: ConfigurationLifeCycle
  onServerReady?: ConfigurationLifeCycle
}

const noop = () => {}

export function createConfiguration(options: ConfigurationOptions) {
  const { onReady, onStop, onConfigLoad, onServerReady } = options
  return createConfigurationFromCore(options)
    .onReady(isFunction(onReady) ? onReady : noop)
    .onStop(isFunction(onStop) ? onStop : noop)
    .onConfigLoad(isFunction(onConfigLoad) ? onConfigLoad : noop)
    .onServerReady(isFunction(onServerReady) ? onServerReady : noop)
}
