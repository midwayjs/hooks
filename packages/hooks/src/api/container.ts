import { __decorate } from 'tslib'
import { Inject, Provide } from '@midwayjs/decorator'
import { ApiFunction } from '@midwayjs/hooks-core'

export type CreateOptions = {
  fn: ApiFunction
  providerId: string
  parseArgs: (inputs: { ctx: any; args: any[] }) => any[]
  handlerDecorators?: any[]
  classDecorators?: any[]
}

export function createFunctionContainer(options: CreateOptions) {
  const {
    fn,
    providerId,
    parseArgs,
    handlerDecorators = [],
    classDecorators = [],
  } = options

  let FunctionContainer = class {
    ctx: any
    async handler(...handlerArgs: any[]) {
      const args = parseArgs({
        ctx: this.ctx,
        args: handlerArgs,
      })
      return await fn(...args)
    }
  }

  Object.defineProperty(FunctionContainer, 'name', {
    value: providerId,
  })

  __decorate([Inject()], FunctionContainer.prototype, 'ctx', void 0)
  __decorate(
    [...handlerDecorators],
    FunctionContainer.prototype,
    'handler',
    null
  )
  FunctionContainer = __decorate(
    [Provide(providerId), ...classDecorators],
    FunctionContainer
  )
  return FunctionContainer
}
