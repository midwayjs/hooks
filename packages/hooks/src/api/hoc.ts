import {
  Api,
  ApiFunction,
  HooksMiddleware,
  Middleware,
  validateArray,
  validateFunction,
} from '@midwayjs/hooks-core'

type Controller = {
  middleware?: HooksMiddleware[]
}

const { deprecate } = require('util')

/**
 * @deprecated Use `Api(Middleware(...middlewares))` instead
 */
export const withController = deprecate(
  <T extends ApiFunction>(controller: Controller, func: T) => {
    if (controller?.middleware !== undefined) {
      validateArray(controller.middleware, 'controller.middleware')
    }
    validateFunction(func, 'func')
    return Api(Middleware(controller.middleware), func)
  },
  'withController is deprecated. Use `Api(Middleware(...middlewares))` instead.'
)

/**
 * @deprecated Use `Api(Middleware(...middlewares))` instead
 */
export const withMiddleware = deprecate(
  <T extends ApiFunction>(middleware: HooksMiddleware[], func: T) => {
    validateArray(middleware, 'middleware')
    validateFunction(func, 'func')
    return Api(Middleware(middleware), func)
  },
  'withMiddleware is deprecated. Use `Api(Middleware(...middlewares))` instead.'
)
