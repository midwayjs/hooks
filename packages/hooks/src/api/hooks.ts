import type { IMidwayContainer } from '@midwayjs/core'
import {
  MidwayConfigService,
  getCurrentAsyncContextManager,
  ASYNC_CONTEXT_KEY,
} from '@midwayjs/core'
import { ERR_INVALID_ARG_TYPE, validateString } from '@midwayjs/hooks-core'

export function useContext<T = any>(): T {
  const ctx = getCurrentAsyncContextManager()
    .active()
    .getValue(ASYNC_CONTEXT_KEY)
  return ctx as T
}

export function useLogger() {
  const ctx = useContext()
  return ctx.logger
}

export function usePlugin(key: string): any {
  validateString(key, 'key')

  const ctx = useContext()
  return ctx.app[key] || ctx[key]
}

export function useInject<T = any>(identifier: new () => T): Promise<T>
export function useInject<T = any>(identifier: string): Promise<any>
export async function useInject(identifier) {
  if (typeof identifier !== 'function' && typeof identifier !== 'string') {
    throw new ERR_INVALID_ARG_TYPE('identifier', 'class, string', identifier)
  }

  const ctx = useContext()
  const requestContext: IMidwayContainer = ctx['requestContext']
  return requestContext.getAsync(identifier)
}

export function useConfig(key?: string) {
  if (key !== undefined) {
    validateString(key, 'key')
  }

  const ctx = useContext()
  const requestContext: IMidwayContainer = ctx['requestContext']
  return requestContext.get(MidwayConfigService).getConfiguration(key)
}
