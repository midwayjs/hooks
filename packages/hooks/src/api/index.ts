export * from './hooks'
export * from './hoc'
export * from './configuration'
export { HooksComponent as hooks } from './component'
export {
  Api,
  All,
  Get,
  Post,
  Put,
  Delete,
  Patch,
  Head,
  Options,
  Query,
  Params,
  Header,
  Middleware,
  Redirect,
  HttpCode,
  SetHeader,
  ContentType,
} from '@midwayjs/hooks-core'
export * from './operator/validate'
export * from './operator/serverless'
export * from './operator/type'
