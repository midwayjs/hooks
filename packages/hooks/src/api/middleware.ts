import { isHooksMiddleware } from '@midwayjs/hooks-core'

export function useHooksMiddleware(mw: (...args: any[]) => any | any) {
  if (!isHooksMiddleware(mw)) return mw

  return (...args: any[]) => {
    const next = args[1]
    return mw(next)
  }
}
