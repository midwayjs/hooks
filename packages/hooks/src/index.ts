// Runtime Api
export * from './api'
export type { ApiConfig } from '@midwayjs/hooks-core'

// User Config
export { defineConfig, UserConfig } from '@midwayjs/hooks-internal'

// For unit test
export type { BaseTrigger, HttpTrigger } from '@midwayjs/hooks-core'
export { getApiTrigger } from '@midwayjs/hooks-core'
