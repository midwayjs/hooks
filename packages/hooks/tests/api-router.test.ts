import { getApiTrigger, HttpTrigger } from '../src'
import { IMidwayKoaApplication } from '@midwayjs/koa'
import { closeApp, createApp, createHttpRequest } from './utils'
import * as index from './fixtures/api-router/src/api/index'
import { args } from '@midwayjs/rpc'
import supertest from 'supertest'

describe('test koa with api router', () => {
  let app: IMidwayKoaApplication

  beforeAll(async () => {
    app = await createApp('api-router')
  })

  afterAll(async () => {
    await closeApp(app)
  })

  function createRequest(api: any): supertest.Test {
    const trigger = getApiTrigger<HttpTrigger>(api)
    const request: supertest.Test = createHttpRequest(app)[
      trigger.method.toLowerCase()
    ](trigger.path)
    request.set('accept', 'application/json')
    return request
  }

  test('api case', async () => {
    const get = await createRequest(index.get)
    expect(get.status).toEqual(200)
    expect(get.text).toEqual('/api/get')

    const post = await createRequest(index.post).send(args('foo'))
    expect(post.status).toEqual(200)
    expect(post.body).toEqual({
      path: '/api/post',
      input: 'foo',
    })
  })

  test('use validator', async () => {
    const { status } = await createRequest(index.post).send(args(false))
    expect(status).toEqual(422)
  })

  test('with middleware', async () => {
    const { header } = await createRequest(index.withMiddleware)
    expect(header.global).toBeTruthy()
    expect(header.module).toBeTruthy()
    expect(header.function).toBeTruthy()
  })

  test('withHttpOperator', async () => {
    const { status, text, header, type } = await createRequest(
      index.withHttpOperator
    )
    expect(text).toEqual('withHttpCode')
    expect(header.from).toEqual('operator')
    expect(header.framework).toEqual('koa')
    expect(type).toEqual('text/html')
    expect(status).toEqual(201)
  })

  test('withRedirectOperator', async () => {
    const { status, text, header } = await createRequest(
      index.withRedirectOperator
    )
    expect(status).toEqual(301)
    expect(header.location).toEqual('/redirect')
    expect(text).toEqual('withRedirectOperator')
  })

  test('withValidateHttp', async () => {
    // Incorrect params format, expect number, but got string
    await createHttpRequest(app)
      .post('/api/withValidateHttp/foo')
      .set('accept', 'application/json')
      .expect(422)

    // Missing Query
    await createHttpRequest(app)
      .post('/api/withValidateHttp/123')
      .set('accept', 'application/json')
      .expect(422)

    // Missing Header
    await createHttpRequest(app)
      .post('/api/withValidateHttp/123')
      .set('accept', 'application/json')
      .query({ isQuery: 'foo' })
      .expect(422)

    // Incorrect data format, expect string, but got number
    await createHttpRequest(app)
      .post('/api/withValidateHttp/123')
      .set('accept', 'application/json')
      .set('isHeader', 'foo')
      .query({ isQuery: 'foo' })
      .send(args(1))
      .expect(422)

    // Valid Case
    const response = await createHttpRequest(app)
      .post('/api/withValidateHttp/123')
      .set('accept', 'application/json')
      .set('isheader', 'foo')
      .query({ isQuery: 'foo' })
      .send(args('Midway'))
      .expect(200)

    expect(response.body).toMatchSnapshot()
  })

  test('slot', async () => {
    const { status, body } = await createHttpRequest(app)
      .get('/slot/withSlot')
      .query({ query: 'query' })
      .set('header', 'header')

    expect(status).toEqual(200)

    expect(body.header.header).toEqual('header')
    expect(body.query.query).toEqual('query')
    expect(body.params.slot).toEqual('slot')
  })
})
