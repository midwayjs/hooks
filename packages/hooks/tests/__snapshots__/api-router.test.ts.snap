// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test koa with api router withValidateHttp 1`] = `
Object {
  "data": "Hello Midway",
  "header": "foo",
  "params": "123",
  "query": "foo",
}
`;
